# Evaluation de la qualité des données
Le but de ce notebook est d'identifier les problèmes de qualité des données. 

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
engine = create_engine('sqlite:///data.db', echo=False)
con = engine.connect()
df_patient = pd.read_sql('select * from patient', con=con)
df_pcr = pd.read_sql('select * from test', con=con)
con.close()

### Analyse rapide des données

In [3]:
# quick data check 
df_patient.head()

patient_id given_name    surname  street_number            address_1  \
0      221958    matisse     clarke           13.0          rene street   
1      771155     joshua     elrick           23.0         andrea place   
2      231932      alice     conboy           35.0     mountain circuit   
3      465838     sienna   craswell           39.0  cumberlegeicrescent   
4      359178     joshua  bastiaans          144.0        lowrie street   

          suburb postcode state  date_of_birth   age phone_number address_2  
0     ellenbrook     2527    wa     19710708.0  32.0  08 86018809  westella  
1   east preston     2074   nsw     19120921.0  34.0  02 97793152   foxdown  
2       prospect     2305   nsw     19810905.0  22.0  02 20403934      None  
3          henty     3620    wa     19840809.0  30.0  02 62832318    jodane  
4  campbell town     4051   nsw     19340430.0  31.0  03 69359594      None

Le Dataframe df_patient contient les informations personnelles sur les patients. Dans le détail, on y trouve les 12 colonnes suivantes qui sont explicites:  
* patient_id : Un  identifiant qui semble représenter un clé	
* given_name	
* surname	
* street_number
* address_1	
* suburb	
* postcode	
* state	
* date_of_birth	
* age	
* phone_number	
* address_2


In [4]:
df_pcr.head()

patient_id       pcr
0      653091         N
1      347667  Negative
2      708481  Positive
3      148825  Negative
4      150081  Negative

Le Dataframe df_pcr contient une colonne patient_id et une colonne pcr qui contient les résultats des tests PCR. On note déjà une première incohérence seulement en regardant les 4 premières lignes de données: les résultats ne sont pas notés de façon consistantes, ainsi les test négatifs sont noté N ou Negative. Nous explorerons ce problème complètement plus tard dans l'analyse. 

## Analyse des données de df_patient

In [5]:
df_patient.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   patient_id     20000 non-null  int64  
 1   given_name     19560 non-null  object 
 2   surname        19575 non-null  object 
 3   street_number  19618 non-null  float64
 4   address_1      19204 non-null  object 
 5   suburb         19788 non-null  object 
 6   postcode       19801 non-null  object 
 7   state          18010 non-null  object 
 8   date_of_birth  17989 non-null  float64
 9   age            16003 non-null  float64
 10  phone_number   19081 non-null  object 
 11  address_2      7893 non-null   object 
dtypes: float64(3), int64(1), object(8)
memory usage: 1.8+ MB


Il y a 20 000 lignes dans le Dataframe. Et il y a des données manquantes pour toutes les colonnes excepté la colonne patient_id. On remarque que les colonnes postcode et phone_number sont représenté sous forme de chaine des caractères et non sous forme de nombres. Par contre date_of_birth est représenté comme un nombre décimal, comme l'age. 
La cellule suivante donne le nombre de valeurs nulles pour chaque colonnes. 


In [6]:
df_patient.isnull().sum().sort_values(ascending=False)

address_2        12107
age               3997
date_of_birth     2011
state             1990
phone_number       919
address_1          796
given_name         440
surname            425
street_number      382
suburb             212
postcode           199
patient_id           0
dtype: int64

La colonne avec le plus de données manquantes est address_2. C'est compréhensible car ce n'est qu'un complément d'adresse qui n'est pas toujours nécessaire. Par contre les données manquantes sont problématiques pour les 2 colonnes qui suivent dans la liste: age et date_of_birth. D'ailleurs ces 2 colonnes sont redondantes: On peut calculer age à partir de date_of_birth. 
En bas de la liste, on trouve patient_id qui n'a pas de valeurs nulles. 


### Analyse statistique des colonnes numériques

In [7]:
df_patient.describe()

patient_id  street_number  date_of_birth           age
count   20000.000000   19618.000000   1.798900e+04  16003.000000
mean   553918.938750      98.812468   1.949638e+07     27.534337
std    260234.961111    2481.908686   2.926916e+05      7.807008
min    100005.000000       0.000000   1.290041e+07      0.000000
25%    330194.250000      10.000000   1.924113e+07     24.000000
50%    556327.500000      24.000000   1.950032e+07     29.000000
75%    779419.500000      60.000000   1.975052e+07     33.000000
max    999993.000000  342951.000000   1.999123e+07     92.000000

De la cellule ci-dessus, nous ne pouvons pas trop tirer de conclusion, excepté pour la colonne age. La moyenne pour la colonne age est 27.534337 et la médianne 29. De plus le 75% percentile est 33. Bien que l'age maximal est 92, nous voyons que la majorité des données sont pour des ages autour de 30 ans. Et donc ces données ne sont pas représentative de toute la population. 

## Colonne patient_id
La colonne patient_id semble être une clé. Et de ce fait toutes les valeurs devraient êtres unique pour identifier correctement une personne. Vérifions ce point. 

In [8]:
# Using the option keep=False to show all the duplicated rows. 
# Moreover a row can be duplicated more than once
df_patient[df_patient.duplicated(subset=['patient_id'], keep=False)].sort_values(by=['patient_id'])

patient_id given_name        surname  street_number  \
12117      109304    zachary      farronato           30.0   
14839      109304     bailey      donaldson           20.0   
4386       110207       toby          brock            4.0   
12989      110207       zali          brock           32.0   
10184      115791     hannah         clarke           70.0   
...           ...        ...            ...            ...   
10507      990695       erin       braunack           49.0   
8764       990936        amy          royle           90.0   
12563      990936   samantha          green           21.0   
2385       994235      trent  stewart-jones          129.0   
7961       994235     teresa      chowdhury           60.0   

                 address_1        suburb postcode state  date_of_birth   age  \
12117       outtrim avenue          como     2196   vic     19090801.0  31.0   
14839       tardent street          ryde     0812   qld     19580310.0  26.0   
4386     merriman crescent      baralaba     3025   nsw     19000424.0  35.0   
12989        hedger street        toorak     5038   act            NaN  22.0   
10184      galmarra street      mayfield     7010   vic     19830828.0  25.0   
...                    ...           ...      ...   ...            ...   ...   
10507     moondarra street   broken hill     2640   qld     19830122.0  30.0   
8764     whittell crescent       coramba     5032    sa     19950326.0   NaN   
12563       brierly street     ardrossan     2140  None     19380210.0  29.0   
2385   macfarland crescent    wangaratta     2732   nsw            NaN   NaN   
7961                  None  chester hill     2147   vic     19010407.0  27.0   

      phone_number       address_2  
12117  07 22894061       the reefs  
14839  07 13479210            None  
4386   08 33842007         leitrim  
12989  08 96818512            None  
10184  04 70760611            None  
...            ...             ...  
10507  03 69523317         yuulong  
8764   08 07309295  tewantin plaza  
12563  02 51600621            None  
2385   07 98662458       mountview  
7961   02 83271460            None  

[403 rows x 12 columns]

Ces 403 lignes dupliqués sont vraiment un problème, car on voit que ce sont des personnes différentes. Les duplications ne viennent pas de fautes de frappes dans la saisie du nom, de l'adresse, ...
Regardons le test PCR pour un de ces patient_id dupliqués. 


In [9]:
# Test result for one of the duplicated patient_id 
df_pcr[ df_pcr['patient_id'] == 109304 ]

patient_id pcr
1960      109304   N

Le test est négatif pour le patient 109304, mais on n'a aucun élément pour dire s'il s'applique à Mr farronato ou Mr donaldson. Donc la solution sera de supprimer tous les patient_id dupliqués dans le dataframe df_patient. 

### Colonne date_of_birth
La colonne date_of_birth est censé représenter la date de naissance, mais le format utilisé est surprenant: C'est un nombre décimal avec 8 chiffres. Cependant lorsque l'on compare la colonne date_of_birth avec la colonne age, il n'y a aucun lien logique. Les premiers patients trié par date_of_birth sont listés dans la cellule suivante.

In [11]:
df_patient.sort_values(by=['date_of_birth']).head(50)

patient_id given_name      surname  street_number  \
8882       258272       jack       jessup          509.0   
11591      986226       lara        white            6.0   
1689       707847       rosa        crook           27.0   
1102       218033     taylah         None            3.0   
11936      726441    timotby         webb          100.0   
18880      400767    timothy         webb           10.0   
13336      981450      rocca     samantha           29.0   
14362      161596   samantha        rocca           29.0   
15408      847075       None        haupt           46.0   
18101      690620      riley        dixon           47.0   
19648      805614       None       nguyen           30.0   
12075      799793      riley        green           36.0   
2205       681273     thomas       marris           47.0   
13090      614080     ridley        green            2.0   
19009      953268     emiily         zilm           10.0   
11023      873931      james           co           13.0   
6500       986788    zachary         lomp            7.0   
6785       847567     joshua        binns           49.0   
14883      249177     kallie         reid            3.0   
14896      205153    melissa         dent            7.0   
2703       751176      flynn       hazell          192.0   
10004      266562     thomas        wynne           17.0   
6677       875292      jenna     campbell           11.0   
16673      791572      grace  schoonwater            5.0   
7004       214456        sam      pennell            2.0   
18277      159263   danielle      haskett            7.0   
14171      406468     harley     mcgregor          225.0   
15427      557864    william     stronach           26.0   
10823      932821      kydan     handberg           16.0   
7101       340011     tahlia      coleman           24.0   
15771      694055     tahlia      coleman           24.0   
561        814998     tahlia      coleman           24.0   
17871      480129     keeley  arulanandam           27.0   
12653      930890    reganne         fysh           41.0   
12124      358490      adela     shepherd           20.0   
15813      840749      elana        white           11.0   
8980       583761       webb      timothy           10.0   
187        424767    madison        ebert           25.0   
16906      445590       zane     campbell            5.0   
1487       357009     joshua     campbell           45.0   
18787      331789     ashlie        white           35.0   
10356      100727       jake       finlay            3.0   
16332      369284       None        jolly           11.0   
15904      115909       liam     dedicoat           58.0   
4354       502980      james         webb           58.0   
14095      880305      lucas  voroniansky           15.0   
9899       972377     stella      redmond           22.0   
12219      307925      dylan       evison           14.0   
10242      222357      jacob     roxburgh            7.0   
14051      297170      aiden         reid            1.0   

                   address_1            suburb postcode state  date_of_birth  \
8882             stone place            batlow     None   qld     12900410.0   
11591     collocott crescent     alice springs     5031  None     19000105.0   
1689          wilkins street             cooma     5125   vic     19000106.0   
1102   ern florence crescent       montmorency     5051   vic     19000112.0   
11936                   None       mount colah     3054   vic     19000118.0   
18880                   None       mount colah     3054   vic     19000118.0   
13336         jacka crescent         edgecilff     6024   nsw     19000119.0   
14362         jacka crescent         edgecliff     6024   nsw     19000119.0   
15408                   None         devonport     4871    wa     19000120.0   
18101    investigator street          hillside     4870    sa     19000121.0   
19648           cleeve place          ballarat     0812   nsw     

La valeur la plus faible est "12900410.0". On s'attendrai à ce que ce soit le patient le plus agé. Or il n'a que 34 ans. 
Il y a aussi de l'incohérence entre le patient 847075 et 161596: le patient 161596 est d'une unité ("19000119.0" contre "19000120.0") plus petite que le patient 847075 mais il est plus agé de 11 ans ! (38 ans contre 27 ans). 
Donc telle quelle nous ne savons pas comment utiliser les informations dans la colonne date_of_birth. Nous pouvons essayer d'analyser un peu plus cette colonne. Logiquement les personnes avec le même date_of_birth devrait avoir le même age. 
 

In [52]:
df_patient[df_patient.duplicated(subset=['date_of_birth', 'age'], keep=False) & 
           df_patient['date_of_birth'].notnull() & df_patient['age'].notnull()
          ].sort_values(by=['age', 'date_of_birth'])

patient_id given_name    surname  street_number        address_1  \
1609       423412    talixsa     brammy           11.0     wtreet james   
2407       821864    talissa     brammy           11.0     james street   
582        876665   madrlixe    beadell          117.0     groom street   
3550       549916   madeline    beadell          117.0     groom street   
3411       792841       None  eglineton           12.0   helemon street   
...           ...        ...        ...            ...              ...   
4971       162433        eva      miles           18.0  collings wtreet   
12288      169409        eva      miles           91.0   collingsstreet   
17936      753197        eva      miles           91.0  collings street   
19313      167261       lars    blunden          101.0   armfield place   
19520      760261       lara    blunden          106.0   armfield place   

                   suburb postcode state  date_of_birth   age phone_number  \
1609          maryborough     5333   vic     19060917.0   7.0  02 77001738   
2407          maryborough     5333   vic     19060917.0   7.0  02 77001738   
582               wyoning     4218   nsw     19230525.0   7.0  08 81656149   
3550              wyoming     4218   nsw     19230525.0   7.0  08 81656149   
3411          surry hills     3216   qlg     19300427.0   7.0  02 49550430   
...                   ...      ...   ...            ...   ...          ...   
4971    eight mile plains     6232   vic     19941210.0  43.0  07 71242229   
12288  eight mioe aplains     6232   vic     19941210.0  43.0  07 71242229   
17936   eight mile plains     6232   vic     19941210.0  43.0  07 71242229   
19313          tarragindi     2031    wa     19080704.0  45.0  04 80689941   
19520          tarragindi     2031    wa     19080704.0  45.0  04 80689941   

             address_2  
1609      tourist park  
2407      tourist park  
582               None  
3550              None  
3411        north road  
...                ...  
4971   rosetta village  
12288  rosetta village  
17936  rosetta village  
19313             None  
19520             None  

[1296 rows x 12 columns]

Mais cela ressemble davantage à des personnes dupliquées. Il serait interresssant de regarder de nouveau ce point, une fois que le dataframe sera nettoyé des doublons.  

### Analyse des doublons
Les colonnes restantes (given_name, surname,  ne surname, street_number, suburb, postcode, state, age, phone_number, et address_2) ne semblent pas poser de problèmes de représentation de données. Sont-elles pour autant fiables? Permettent-elles d'identifier une personne? Y-il-a t'il des doublons dues à des fautes de frappe dans la saisie des information? 
La colonne susceptible de plus précisément identifier une personne est phone_number. 
Regardons s'il y a des lignes dupliquées pour cette colonne. 

In [37]:
df_patient[df_patient.duplicated(subset='phone_number', keep=False)].sort_values(by=['phone_number'])

patient_id given_name        surname  street_number  \
2634       311830     taaila           None           16.0   
3872       210155     taalia           None           19.0   
16605      525466     taalia           None           19.0   
248        870746      tyler           able           15.0   
4986       268616      tyler           able          154.0   
...           ...        ...            ...            ...   
19899      668176      holly        bristow           35.0   
19901      629087       None  stewart-jones           20.0   
19922      563445     rachel    shepherdson            9.0   
19932      859893       ruby      nascivera           35.0   
19966      234845       kyle         scrbak           83.0   

                  address_1         suburb postcode state  date_of_birth  \
2634         mcnicoll place         orange     6230   vic     19180603.0   
3872         mcnicollyplace         orange     6230   vic     19180603.0   
16605        mcnicoll place         orange     6230   vic     19180603.0   
248    marcus clarke street       wannissa     3333   nsw     19830504.0   
4986   marcus clarke street      wanniassa     3333   nsw     19830504.0   
...                     ...            ...      ...   ...            ...   
19899          ebden street  dangar island     5024   vic     19501102.0   
19901         jonsson court         aspley     3215   nsw     19490912.0   
19922       calder crescent          junee     6163  None     19151122.0   
19932          heard street      roselands     4814   nsw     19841221.0   
19966        wanderer court     maida vale     3068   qld     19560328.0   

        age phone_number           address_2  
2634    NaN  02 00325977                None  
3872   64.0  02 00325977                None  
16605  26.0  02 00325977                None  
248     NaN  02 00493853                None  
4986    NaN  02 00493853                None  
...     ...          ...                 ...  
19899  33.0         None          rocky flat  
19901  36.0         None                None  
19922  31.0         None             fernlea  
19932  36.0         None       sheep station  
19966  39.0         None  little springhurst  

[2771 rows x 12 columns]

On voit qu'il y 2771 lignes dupliquées. On note que les 3 premières lignes semblent représenter le même patient. Mais il y a des faute de frappe sur address_1 ou street_number. Le bas de la liste montre également que toutes les valeurs nulles sont considérées comme dupliquées, ce qui n'est pas ces patients qui sont clairement différents. Enlevons les valeurs nulles. 

In [38]:
df_patient[df_patient.duplicated(subset='phone_number', keep=False) &
          df_patient['phone_number'].notnull()].sort_values(by=['phone_number'])

patient_id given_name     surname  street_number             address_1  \
2634       311830     taaila        None           16.0        mcnicoll place   
3872       210155     taalia        None           19.0        mcnicollyplace   
16605      525466     taalia        None           19.0        mcnicoll place   
248        870746      tyler        able           15.0  marcus clarke street   
4986       268616      tyler        able          154.0  marcus clarke street   
...           ...        ...         ...            ...                   ...   
7100       178533       ryab  samakovlis           70.0         ferrier place   
16577      641889       ryan  samakovlis           70.0         ferrier place   
5195       912481      kiera       feast            5.0        bateman street   
18207      570917      kierl       feqsa            5.0        bateman street   
13964      904253      kiera       feawt            4.0                  None   

          suburb postcode state  date_of_birth   age phone_number  \
2634      orange     6230   vic     19180603.0   NaN  02 00325977   
3872      orange     6230   vic     19180603.0  64.0  02 00325977   
16605     orange     6230   vic     19180603.0  26.0  02 00325977   
248     wannissa     3333   nsw     19830504.0   NaN  02 00493853   
4986   wanniassa     3333   nsw     19830504.0   NaN  02 00493853   
...          ...      ...   ...            ...   ...          ...   
7100      kirwan     7301   nsw            NaN  31.0  08 98483679   
16577     kirwan     7310   nsw            NaN  31.0  08 98483679   
5195         moe     5125   nsw            NaN  33.0  08 99975497   
18207        moe     5125   nsw            NaN  33.0  08 99975497   
13964        moe     5125   nsw            NaN  33.0  08 99975497   

              address_2  
2634               None  
3872               None  
16605              None  
248                None  
4986               None  
...                 ...  
7100              miami  
16577             miami  
5195   workmans cottage  
18207  workmans cottage  
13964  workmans cottage  

[1852 rows x 12 columns]

Nous avons maintenant 1852 lignes dupliquées. Cet extrait nous montre aussi que les fautes de frappe peuvent apparaître dans n'importe quelle colonne: given_name, surname, ...



In [42]:
df_patient[df_patient.duplicated(subset=['given_name','surname', 'address_1'], keep=False) &
          df_patient[['given_name','surname', 'address_1']].notnull().any(axis=1)].sort_values(
          by=['given_name','surname', 'address_1'])

patient_id given_name     surname  street_number         address_1  \
16951      592646      abbey      cadman            5.0         mugga way   
17258      593406      abbey      cadman            5.0         mugga way   
2781       959414      abbey     cropper           17.0   bowden crescent   
5089       560115      abbey     cropper           16.0   bowden crescent   
6627       961858      abbey       green            5.0    wilkins street   
...           ...        ...         ...            ...               ...   
16552      279300       None       polak           35.0        jess place   
1775       310017       None        wehr           57.0  christmas street   
9140       847320       None        wehr           57.0  christmas street   
7968       215375       None  zimmermann          447.0   birdwood street   
10591      749179       None  zimmermann          447.0   birdwood street   

               suburb postcode state  date_of_birth   age phone_number  \
16951      burpengary     2554  None     19920502.0  29.0  03 45996223   
17258      burpengary     2545  None     19920502.0  29.0  03 45969923   
2781         ringwood     6237   qld     19200312.0  29.0  02 90211615   
5089         ringwood     6231   qld     19200312.0  21.0  02 90211615   
6627          ballina     6110  None     19040209.0   7.0  08 75306674   
...               ...      ...   ...            ...   ...          ...   
16552       annandale     2250   qld     19531130.0   NaN  03 33180608   
1775   mount victoria     3068   nsw     19820227.0  33.0  03 15451529   
9140   mount victoria     3068  None     19829327.0  33.0  03 15451529   
7968       beerburrum     4221   vic     19724011.0  29.0  03 10366613   
10591      beerburrum     4221   vic     19721001.0  29.0  03 10366613   

         address_2  
16951         None  
17258         None  
2781       pangani  
5089       pangani  
6627          None  
...            ...  
16552  bonnie doon  
1775          None  
9140          None  
7968     wildefell  
10591         None  

[932 rows x 12 columns]

L'exemple ci-dessus montre la difficulté de détecter les doublons lorsque l'on ne peut pas s'appuyer sur un idetificateur fort comme phone_number: On a besoin de 3 colonnes ('given_name','surname', 'address_1') pour commencer à identifier un patient. Et l'exemple ci-dessus ne tient pas contre du fait qu'il peut y avoir des fautes de frappe sur ces trois colonnes. 
On voit également qu'il peut y avoir des fautes de frappe sur phone_number.
On note également que la colonne date_of_birth semble être un identificateur de patient, même s'il nous n'avons pas la clé pour récupérer la date de naissance. Donc cette colonne sera utile pour la suppression des doublons. 

In [45]:
df_patient[df_patient.duplicated(subset=['date_of_birth','surname'], keep=False) &
          df_patient[['date_of_birth']].notnull().any(axis=1)].sort_values(
          by=['date_of_birth','surname'])

patient_id given_name    surname  street_number           address_1  \
11936      726441    timotby       webb          100.0                None   
18880      400767    timothy       webb           10.0                None   
561        814998     tahlia    coleman           24.0        furey street   
7101       340011     tahlia    coleman           24.0        furey street   
15771      694055     tahlia    coleman           24.0         furey steet   
...           ...        ...        ...            ...                 ...   
16782      193878        amy      mason           12.0  balsillie crescent   
3976       810644    juliana  grosvenor            5.0       connelly pace   
11283      793044    juliana  grosvenor            5.0      connelly place   
15581      697565    juliana  grosvenor            2.0      connellt place   
18365      692127    juliqna  grosvenor            1.0      connelly place   

                     suburb              postcode state  date_of_birth   age  \
11936           mount colah                  3054   vic     19000118.0   NaN   
18880           mount colah                  3054   vic     19000118.0   NaN   
561                boondall                  3163   nsw     19000310.0  24.0   
7101               boondall                  3163   nsw     19000310.0  19.0   
15771              boondall                  3163   nsw     19000310.0  19.0   
...                     ...                   ...   ...            ...   ...   
16782             view bank                  2126   vic     19991025.0  12.0   
3976                   3023  port noarlunga south   tas     19991215.0   NaN   
11283  port noarlunga south                  3023   tas     19991215.0   NaN   
15581  port noarlunha south                  3023   tas     19991215.0   NaN   
18365  port noarlunga south                  3023   tad     19991215.0   NaN   

      phone_number address_2  
11936  04 40675254      None  
18880  04 40675254      None  
561    02 02079822      None  
7101   02 02079822      None  
15771  02 32161703      None  
...            ...       ...  
16782  02 48437211      None  
3976   03 55227740      None  
11283  03 55227740      None  
15581  03 55227740      None  
18365  03 55227740      None  

[1148 rows x 12 columns]

## Analyse des données de df_pcr
Nous allons d'abord vérifier s'il y a des doublons pour la colonne ptient_id

In [48]:
df_pcr[df_pcr.duplicated(subset=['patient_id'], keep=False)].sort_values(by=['patient_id'])

patient_id       pcr
169       159500         N
2934      159500         N
8469      169519  Positive
1842      169519  Negative
5618      212217         N
...          ...       ...
5542      887646  Negative
1025      975512  Negative
2331      975512         N
7477      990695         N
3407      990695  Positive

[70 rows x 2 columns]

On voit qu'il y a deux cas de figure pour les doublons:  
* Soit la valeur pcr est identique: Il faut dans ce cas juste supprimer le ligne en double 
* Soit les valeur pcr sont différente: Dans ce cas on ne peut pas savoir laquelle est bonne. Il faut donc supprimer les 2 lignes